### DISCUSSION TITLE
How to copy the file when the project is build from cake script ?

### DISCUSSION THREAD

**Discussion by Vijaykumarj** [cake-build/cake#3832] [Q&A] [answered]
Hi,

     I am a newbie in cake build and prepared a cake script for the deployment of my ASP.Net Core 5 project and I am using Cake 2.1.0. 
 Do we have any feature like "Build Actions in Visual Studio - i.e Copy the files to output directory" using Cake build?

 Ref Link: https://docs.microsoft.com/en-us/visualstudio/ide/build-actions?view=vs-2022

  Any suggestions or recommendations on this.

   Thanks in advance.

**Comment by nils-a** [comment, marked as answer]
@Vijaykumarj, There are a lot of file operations available. Take a look at the ["copy" section of the "file-operations" documentation](https://cakebuild.net/dsl/file-operations/#Copy)

**Reply by Vijaykumarj** [reply]
Thank you nils-a, We can use this approach by copying the files from the source to the destination folder. But I have to copy the files or folders which are marked as "Copy to Output Directory" as "Content". So to achieve this, we need to traverse the csproj file for which files need to copy to publish folder. This is the tricky part to identify in csproj file in the cake build script. 

Do we have any better approach to achieve this? Please suggest this.

**Reply by nils-a** [reply]
@Vijaykumarj, I would suggest building the csproj, in which case all build steps (including "Copy to Output Directory") of the project will be executed. 
Depending on the project, you can use either [`MSBuild()`](https://cakebuild.net/api/Cake.Common.Tools.MSBuild/MSBuildAliases/) or [`DotNetBuild()`](https://cakebuild.net/api/Cake.Common.Tools.DotNet/DotNetAliases/)

**Reply by Vijaykumarj** [reply]
@nils-a , I am using DotNetBuild and my Post Build Event is like below:   

![image](https://user-images.githubusercontent.com/26544969/157008806-c6f38aa5-202c-443e-9e9d-768137803b62.png)


But I got the error as copy "***Undefined** *\Lib\*.*" while running from my cake build script.  
How to pass the value of "SolutionDir" from Cake build Script?

**Reply by nils-a** [reply]
This is not really Cake specific, the same happens when you manually build the project instead of the solution. 

If you build the solution instead of the project, `SolutionDir` will be automatically set. Alternatively, [`DotNetBuild`](https://cakebuild.net/api/Cake.Common.Tools.DotNet/DotNetAliases/) has an overload that takes [`DotNetBuildSettings`](https://cakebuild.net/api/Cake.Common.Tools.DotNet.Build/DotNetBuildSettings/). These settings have a property `MSBuildSettings` of type [`DotNetMSBuildSettings`](https://cakebuild.net/api/Cake.Common.Tools.DotNet.MSBuild/DotNetMSBuildSettings/). There you can add your properties to the `Properties` property.

**Reply by Vijaykumarj** [reply]
@nils-a It's true, but getting the below error 
![image](https://user-images.githubusercontent.com/26544969/157030187-9fcff4d5-c3c9-4537-b774-eb15aaa25b6c.png)

when we set the properties to [DotNetMSBuildSettings](https://cakebuild.net/api/Cake.Common.Tools.DotNet.MSBuild/DotNetMSBuildSettings/)

![image](https://user-images.githubusercontent.com/26544969/157030292-228c36d8-734d-4a40-976d-46db8edde395.png)

Am I missing anything?

**Reply by nils-a** [reply]
Yes, the `Properties` property can not be assigned. It is read-only. Try modifying it like
```cs
var msbuildSettings = new DotNetMSBuildSettings() {
  // Verbosity & Env...
};
msbuildSettings.Properties.Add("SolutionDir", "QA");
msbuildSettings.Properties.Add("TargetDir", "ABCD");
// foreach....
```

**Reply by Vijaykumarj** [reply]
HI @nils-a  I made changes as you suggested but getting below error due to 


`Properties | IDictionary<string, ICollection<string>>`


![image](https://user-images.githubusercontent.com/26544969/157037282-8abdd06a-90c8-4a05-b3bb-156b6d3c51f3.png)

**Reply by nils-a** [reply]
Sorry, typed from memory. The second parameter of the `Add()` method needs to be an `ICollection<string>`, so the correct code would for example be `msbuildSettings.Properties.Add("SolutionDir", new[]{"QA"});`.

**Reply by Vijaykumarj** [reply]
Hi, @nils-a  I was passed the SolutionDir property with the value to msbuildSettings even though getting the error as "Undefined*" while running the PostBuild event.

**Reply by nils-a** [reply]
@Vijaykumarj could you run Cake with the parameter `--verbosity=diagnostic` ?

There should be some line similar to 
```txt
Executing: "C:/Program Files/dotnet/dotnet.exe" build "./foo/foo.csproj" /property:SolutionDir=c:\
```

Could you post that line?

**Reply by Vijaykumarj** [reply]
@nils-a I ran the Cake script with parameter `--verbosity=diagnostic `. Properties are passed with DotNetBuild (Projects). But I am not able to pass the Properties to my Unit Tests project with DotNetCoreTestSettings. This Unit Test project has referenced one of the projects which contain PostBuild Event.

![image](https://user-images.githubusercontent.com/26544969/157244757-36e43412-9844-433e-904e-30d294df4911.png)


 
Code Snippet for Run Unit Test project: 

`Task("Run-Unit-Tests")
    .IsDependentOn("Copy webconfig")
    .Does(() =>
{
    Information("project.FullPath");
    var msbuildSettings = new DotNetMSBuildSettings() {
            Verbosity = DotNetVerbosity.Minimal,            
        };
    msbuildSettings.Properties.Add("SolutionDir", new[]{solutionDir});
    foreach(var project in projects.Where(p => p.IsTestProject))
    {
       var buildSettings = new DotNetCoreTestSettings()
       {
          Configuration = configuration,
          MSBuildSettings = msbuildSettings
       };
       DotNetTest(project.FullPath, buildSettings);
	   Information(project.FullPath);
    }
`

There is no property or msbuildsettings pass in DotNetCoreTestSettings as per official documentation. 

Can you guide me on how to pass the Properties to DotNetTest?

**Reply by nils-a** [reply]
@Vijaykumarj sure. `DotNetBuild` the test project, as shown above. 
Then call `DotNetTest` with [`DotNetTestSettings.NoBuild`](https://cakebuild.net/api/Cake.Common.Tools.DotNet.Test/DotNetTestSettings/) set to `true`.

**Reply by Vijaykumarj** [reply]
thanks, @nils-a, It's working as expected. Thank you so much. I appreciate your support.
